hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerStartupCache.java

```diff
@@ -0,0 +1,366 @@
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Preconditions;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.hadoop.hdds.conf.ConfigurationSource;
+import org.apache.hadoop.hdds.utils.db.RDBStore;
+import org.apache.hadoop.ozone.OzoneConsts;
+import org.apache.hadoop.ozone.container.common.interfaces.Container;
+import org.apache.hadoop.ozone.container.common.utils.DatanodeStoreCache;
+import org.apache.hadoop.ozone.container.common.volume.HddsVolume;
+import org.apache.hadoop.ozone.container.keyvalue.KeyValueContainer;
+import org.apache.hadoop.ozone.container.keyvalue.KeyValueContainerData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Container YAML cache for reducing random IO when reading Container YAML files.
+ */
+public class ContainerStartupCache {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ContainerStartupCache.class);
+  private static final String CACHE_FILE_NAME = "container_cache.json";
+  private static final String CACHE_FILE_TEMP_SUFFIX = ".tmp";
+  private static final String CHECKSUM_PLACEHOLDER = "CHECKSUM_PLACEHOLDER";
+
+  private final ObjectMapper objectMapper;
+
+  /**
+   * Versioning for Container Cache format.
+   */
+  public enum CacheVersion {
+    INITIAL_VERSION(1, "Initial container cache format"),
+
+    FUTURE_VERSION(-1, "Used internally when an unknown cache version is encountered");
+
+    public static final CacheVersion CURRENT = latest();
+    public static final int CURRENT_VERSION = CURRENT.version;
+
+    private final int version;
+    private final String description;
+
+    CacheVersion(int version, String description) {
+      this.version = version;
+      this.description = description;
+    }
+
+    public String description() {
+      return description;
+    }
+
+    public static CacheVersion fromValue(int value) {
+      for (CacheVersion v : values()) {
+        if (v.version == value) {
+          return v;
+        }
+      }
+      return FUTURE_VERSION;
+    }
+
+    private static CacheVersion latest() {
+      CacheVersion[] versions = CacheVersion.values();
+      // Return the latest version (excluding FUTURE_VERSION which is always last)
+      return versions[versions.length - 2];
+    }
+  }
+
+  public ContainerStartupCache() {
+    this.objectMapper = new ObjectMapper();
+    this.objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+  }
+
+  /**
+   * Cache metadata for containers in the volume.
+   */
+  public static class CacheMetadata {
+    private Map<Long, String> containerYamlData;
+    private long rocksDbSequenceNumber;
```

Comment:
It might be risky to use rocksDbSequenceNumber, as,
- threads will be running for delete blocks, and other processes. So might be difficult to ensure db is not modified during shutdown and when this being updated.
- rocksdb can change offline while fixing issue OR over future requirement like upgrade, and other startup logic.

So need remove dependency with rockdb internal part to do consistency check.

Reply:
@xichen01 , thanks for proposing this improvement. 
I understand we need an anchor as the indicator that whether the cached file is update-to-date and can be used, or stale so that it should not be referred. While this is a yaml file cache, even RocksDB rocksDbSequenceNumber has changed, for example, some compactions, it doesn't indicate that the yaml file on disk are changed too so that this cache file becomes stale. 

@xichen01 , is this adopted in real cluster?  Have we got any info about the cache hit ratio, and miss ratio given this implementation?

Reply:
> thanks for proposing this improvement.
I understand we need an anchor as the indicator that whether the cached file is update-to-date and can be used ...

Using the rocksDbSequenceNumber is a stricter indicator, which may cause a usable, latest cache be dropped, but cannot cause a stale cache be loaded. I have not find a more accurate indicator yet.